### PR TITLE
[Main] [all-e]CompAuxNum and CompAuxLib columns should be informed for Payment Discount lines in the French Audit File.Initial Commit

### PR DESCRIPTION
**Pull request by Shikhverma**
[Bug 643980](https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/643980): [MAster] [all-e]CompAuxNum and CompAuxLib columns should be informed for Payment Discount lines in the French Audit File.

Fixes [AB#643980](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/643980)

ISSUE: CompAuxNum and CompAuxLib columns should be informed for Payment Discount lines in the French Audit File.

CAUSE: value not inserted for CompAuxNum and CompAuxLib columns.

SOLUTION: Code added for the CompAuxNum and CompAuxLib columns.

### CODE REVIEW

**Review by v-ankitgoyal — DISMISSED** [review]

File: src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al [hunk 1]
```diff
@@ -672,6 +678,78 @@ codeunit 10826 "Generate File FEC"
...
+    local procedure IsPaymentDiscountAccount(GLAccountNo: Code[20]): Boolean
+    var
+        GeneralPostingSetup: Record "General Posting Setup";
+        CustomerPostingGroup: Record "Customer Posting Group";
+        VendorPostingGroup: Record "Vendor Posting Group";
+    begin
+        if GLAccountNo = '' then
+            exit(false);
+
+        if not PmtDiscountAccountsInitialized then begin
```
Comment (AleksanderGladkov) [inline src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al]: @Shikhverma I would put this whole block in a new function, like CollectPmtDiscountAccounts or InitializePmtDiscountAccountDict / InitPmtDiscountAccountList etc.

```
procedure InitPmtDiscountAccountList()
if not PmtDiscountAccountsInitialized then begin
...
end;
```

```
procedure IsPaymentDiscountAccount(...)
if GLAccountNo = '' then exit;

InitPmtDiscountAccountList();
exit(PmtDiscountAccounts.ContainsKey(GLAccountNo));
```

File: src/Apps/FR/FECAuditFile/test/src/FECAuditFileExportTests.Codeunit.al [hunk 2]
```diff
@@ -2387,6 +2444,24 @@ codeunit 148017 "FEC Audit File Export Tests"
...
 
+    local procedure FindPaymentDiscountLineWithParty(var FileInStream: InStream; PmtDiscAccountNo: Code[20]; ExpectedPartyNo: Code[20]; ExpectedPartyName: Text): Boolean
+    var
+        LineFields: List of [Text];
+        LineToRead: Text;
+    begin
+        while not FileInStream.EOS() do begin
+            FileInStream.ReadText(LineToRead);
+            LineFields := LineToRead.Split('|');
+            if LineFields.Count() >= 8 then
```
Comment (AleksanderGladkov) [inline src/Apps/FR/FECAuditFile/test/src/FECAuditFileExportTests.Codeunit.al]: @Shikhverma is it possible not to use conditions and find the line that you needed?

File: src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al (hunk 1 above)
Comment (Shikhverma) [reply src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al]: > @Shikhverma I would put this whole block in a new function, like CollectPmtDiscountAccounts or InitializePmtDiscountAccountDict / InitPmtDiscountAccountList etc.
> 
> ```
> procedure InitPmtDiscountAccountList()
> if not PmtDiscountAccountsInitialized then begin
> ...
> end;
> ```
> 
> ```
> procedure IsPaymentDiscountAccount(...)
> if GLAccountNo = '' then exit;
> 
> InitPmtDiscountAccountList();
> exit(PmtDiscountAccounts.ContainsKey(GLAccountNo));
> ```

Updated

**Review by v-ankitgoyal — DISMISSED** [review]

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

This PR adds customer or vendor information to FEC export lines whose G/L account is configured as a payment discount account. The linked bug is a valid FEC scenario: the payment discount line was missing CompAuxNum and CompAuxLib while the related customer lines had them.

The fix is targeted. It keeps the existing receivables, payables, bank, and multiple-posting-group logic, then fills the party only when the line still has no party and the account is a payment discount account. The added customer regression test covers the reported path, and there is no public API or BaseApp event dependency in this change.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** - *Add vendor payment discount coverage*
The production code also handles vendor and purchase payment discount accounts. Please add a small vendor test too, so the Vendor branch is protected and future changes cannot break purchase payment discount FEC lines.

**S2** - *Avoid broad setup scans during export*
InitPmtDiscountAccountList() reads all General Posting Setup, Customer Posting Group, and Vendor Posting Group records the first time this check runs. Consider caching by posting group or source account as needed, so large companies do not pay for full setup scans on each export.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed area is financially and legally sensitive because it affects the French audit file. The behavior change is narrow and does not change public signatures, but it can affect any customer or vendor G/L entry posted to an account configured as a payment discount account. The main residual risks are the untested vendor path and the extra setup reads during export.

**Necessity:** The linked Bug 643980 has concrete repro steps and shows a real compliance gap in exported payment discount lines. Without this change, users must manually fix or explain missing auxiliary account data in the FEC file. The scope is appropriate for a bug fix.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9710 round=1 by=alexei-dobriansky at=2026-08-03T10:19:54Z lastSha=6fe92f9cb795e6b848dcd5b8efb034da20ea1427 suggestions=S1,S2

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 2

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The new commit adds vendor payment discount coverage and replaces the full setup-table scans with lazy caches for the posting groups seen during export. Both previous suggestions are addressed in shape.

The vendor test covers the purchase path. The lazy lookup avoids full-table scans, but the result is stored in one global account dictionary. This can make one posting group or source type affect the decision for a later G/L entry, so the export can add customer or vendor data to a line that is not a payment discount line for that entry's setup.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S1 | Add vendor payment discount coverage | Addressed | Vendor regression test added in `f3d259e` (`FECAuditFileExportTests.Codeunit.al:2411-2472`). |
| S2 | Avoid broad setup scans during export | Addressed | Full-table scans were replaced with lazy posting-group and General Posting Setup lookups in `f3d259e`. No separate reply was posted. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

**S3** - *Keep payment discount checks setup-specific*
`GenerateFileFEC.Codeunit.al:716-719` checks one global account dictionary, while lines 735-747 and 767-770 add accounts from different posting groups and from both sales and purchase setup fields. Keep the cached result scoped to the current source type and posting setup, and add a negative test proving that an account cached for one setup does not mark a later unrelated line as a payment discount line.

**S4** - *Cache party details inside the export loop*
`GenerateFileFEC.Codeunit.al:242-243` now calls `SetPartyForPaymentDiscount` for every unresolved party, and lines 693 and 702 read Customer or Vendor each time. Cache the party name and posting group by source type and source number, so a large export does not repeat the same database reads.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** This is a legally sensitive export. The new global account cache can leak payment discount classification across posting groups or between sales and purchase setup fields, which can write incorrect CompAuxNum and CompAuxLib values. The new party lookup also increases database reads for large exports. The FR build and test checks completed successfully, but they do not cover the cross-setup negative case.

**Necessity:** The change remains necessary because payment discount lines need customer or vendor data in the French audit file. The vendor test and lazy setup reads are useful, but the account decision must be tied to the current G/L entry's setup before merge.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9710 round=2 by=alexei-dobriansky at=2026-08-04T11:00:00Z lastSha=f3d259efd9d2a742b3b4ce706c291d1d3ae73f7a suggestions=S1:addressed,S2:addressed,S3:new,S4:new parentRound=1

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 3

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The new commit scopes the payment-discount account cache by posting setup, adds customer/vendor party caches, and adds a negative test for a shared account across customer posting groups. This addresses the N+1 lookup concern and part of the previous setup-leak concern.

One correctness gap remains. The General Posting Setup cache still stores Sales and Purch. payment discount accounts in the same scope, so a Customer entry can match a Purch. payment discount account, and a Vendor entry can match a Sales payment discount account. That can populate CompAuxNum and CompAuxLib on a line that should stay blank.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S1 | Add vendor payment discount coverage | Addressed | Vendor coverage remains present. |
| S2 | Avoid broad setup scans during export | Addressed | The setup scans were replaced by lazy scoped caches. |
| S3 | Keep payment discount checks setup-specific | Not addressed | The cache is posting-setup scoped now, but Sales and Purch. General Posting Setup accounts are still mixed in one scope. |
| S4 | Cache party details inside the export loop | Addressed | Customer and vendor name/posting-group data is now cached by source number. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

**S3** - *Separate sales and purchase discount accounts*
`CacheGenPostingSetupPmtDiscountAccounts` adds Sales and Purch. payment discount accounts to the same scope. For a Customer entry, only Sales Pmt. Disc. accounts should qualify; for a Vendor entry, only Purch. Pmt. Disc. accounts should qualify. Include the source type in this setup scope and add a negative test for the opposite-side account.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** This is a legally sensitive French FEC export. The remaining issue can write customer or vendor auxiliary data to lines that are not payment discount lines for that source type. There is no public API or BaseApp event dependency in this change.

**Necessity:** Bug 643980 is valid and important because payment discount lines need auxiliary account data in the FEC file. The fix is needed, but the account qualification must stay source-specific before merge.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9710 round=3 by=alexei-dobriansky at=2026-08-05T15:51:56Z lastSha=6cd23ae1d61f711e5bc7812df17b12bd31ff50ef reviewKey=18835125094fbbcdd4998a39a3d1477efaf57f946a578729194ef607d4358299 suggestions=S3@dd25e5a2:notaddressed parentRound=2

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 4

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

This PR populates CompAuxNum and CompAuxLib on payment discount lines in the French FEC audit file. The new commit adds `SourceType` to the General Posting Setup cache scope and to the `case` branch that selects which accounts qualify, so a Customer entry only matches Sales payment discount accounts and a Vendor entry only matches Purchase ones. A negative test confirms that a Purchase discount account does not flag a Customer line.

The S3 correctness concern from round 3 is fully resolved. The scope key and the account selection are now aligned: the key includes the source type, and the case block adds only the accounts that belong to that side. Both branches use the same pattern, so the fix is symmetric even though only the Customer-side negative test was added.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S3 | Separate sales and purchase discount accounts | Addressed | `GenPostingSetupScopeKey` and `CacheGenPostingSetupPmtDiscountAccounts` now include `SourceType`; a `case SourceType of` block adds only Sales or Purchase accounts per entry side. Negative test `PurchPmtDiscAccountDoesNotQualifyForCustomerLine` confirms the boundary. |
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The change is limited to the FEC export engine. The scope key change means that a General Posting Setup entry that was previously cached once is now cached separately per source type, which slightly increases dictionary size but eliminates the cross-side contamination risk. No public API or BaseApp event dependency is involved.

**Necessity:** Bug 643980 is a valid compliance gap and the fix is correctly scoped. All earlier concerns have been addressed across the review rounds.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9710 round=4 by=alexei-dobriansky at=2026-08-06T13:14:16Z lastSha=f5749fdda6a41528d56c0b09d53c5d2aa086dd8c reviewKey=a62db34fefeba18f22e51697d8554efc9869e28a51751f284a0ed528ff14f5cf suggestions=S3@dd25e5a2:addressed parentRound=3

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 5

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The single new commit ("Fix Automation") enables `Adjust for Payment Disc.` in General Ledger Setup inside `PurchPmtDiscAccountDoesNotQualifyForCustomerLine`. This was needed to make the payment posting path actually generate a payment discount G/L entry, so the negative test can verify that a Purchase payment discount account does not flag a Customer line.

The fix is correct in intent. The export-engine changes from earlier rounds remain intact and the correctness concern from Round 3 (source-type separation in the account cache) is fully resolved.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S3 | Separate sales and purchase discount accounts | Addressed | `GenPostingSetupScopeKey` and `CacheGenPostingSetupPmtDiscountAccounts` include `SourceType`; negative test `PurchPmtDiscAccountDoesNotQualifyForCustomerLine` confirms the boundary. Marked Addressed in Round 4. |
</details>

<details>
<summary><strong>New observations (commits since round 4)</strong></summary>

**S5** - *Save General Ledger Setup before modifying it in the test*
`PurchPmtDiscAccountDoesNotQualifyForCustomerLine` sets `"Adjust for Payment Disc."` to `true` and calls `Modify(true)`, but it does not call `LibrarySetupStorage.Save(Database::"General Ledger Setup")` first. The `Initialize()` procedure restores only tables that were previously saved, so this change persists across tests and can affect the behavior of any test that runs after this one. Add `LibrarySetupStorage.Save(Database::"General Ledger Setup")` immediately before the `Get()` call.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The export-engine change is legally sensitive (French FEC audit file) but was already reviewed and accepted in Round 4. The only new code is in the test file. The missing `LibrarySetupStorage.Save` is a test isolation risk: in practice, test execution order in AL is not guaranteed, and a later test that expects `Adjust for Payment Disc.` to be false may fail or behave unexpectedly.

**Necessity:** The automation fix is needed to make the negative test exercise the payment posting path. The one-line save call is a low-cost improvement that eliminates the isolation risk.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9710 round=5 by=alexei-dobriansky at=2026-08-07T12:14:36Z lastSha=47e416eb30aeeffa330a3d20e298bf212c709a31 reviewKey=af7ce78e5fa8203c850e8e4d36f4bff1742a1ec7ff9098d36ede195d44fa1675 suggestions=S3@dd25e5a2:addressed,S5@48fa4ad6:new parentRound=4

File: src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al (hunk 1 above)
Comment (Shikhverma) [reply src/Apps/FR/FECAuditFile/app/src/ExportEngine/GenerateFileFEC.Codeunit.al]: Update the changes please review.

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 6

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

This PR populates CompAuxNum and CompAuxLib on payment discount lines in the French FEC audit file by looking up customer or vendor party data when the G/L account is a payment discount account, scoped by source type and posting setup.

The single commit in this round adds `LibrarySetupStorage.Save(Database::"General Ledger Setup")` immediately before the `GeneralLedgerSetup.Get()` call in `PurchPmtDiscAccountDoesNotQualifyForCustomerLine`. This is the correct placement. The `Initialize()` procedure will now restore the General Ledger Setup after the test, so modifying `"Adjust for Payment Disc."` in this test cannot affect tests that run after it.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|---|---|---|---|
| S3 | Separate sales and purchase discount accounts | Addressed | `GenPostingSetupScopeKey` and `CacheGenPostingSetupPmtDiscountAccounts` include `SourceType`; the `case SourceType of` block adds only Sales or Purchase accounts per entry side. Marked Addressed in Round 4. |
| S5 | Save General Ledger Setup before modifying it in the test | Addressed | `LibrarySetupStorage.Save(Database::"General Ledger Setup")` added before `GeneralLedgerSetup.Get()` in `PurchPmtDiscAccountDoesNotQualifyForCustomerLine`. |
</details>

<details>
<summary><strong>New observations (commits since round 5)</strong></summary>

*None — the single commit addresses S5 and introduces no new concerns.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The change in this round is limited to a single line in the test file. All production code was reviewed and accepted in prior rounds. The test isolation risk from round 5 is now closed. No public API or BaseApp event dependency is involved.

**Necessity:** Bug 643980 is a valid compliance gap in the French FEC audit file. All concerns raised across six rounds have been addressed. The fix is correctly scoped and the test suite now covers positive, negative, and isolation cases.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=9710 round=6 by=alexei-dobriansky at=2026-08-10T08:30:12Z lastSha=9c61bd4b8be4b6e74922a8d38adb2afd4c670b96 reviewKey=d7765b4bc2317dbd73ae7f25dd3766e62d8e94fa949eb8befb61f153d055e204 suggestions=S3@dd25e5a2:addressed,S5@48fa4ad6:addressed parentRound=5

**Review by v-ankitgoyal — APPROVED** [review]

**Review by AndersLarsenMicrosoft — APPROVED** [review]

**Review by AleksanderGladkov — APPROVED** [review]
